File: hej/apps/hej-web/lib/api/task-setup.ts
/**
 * Task Setup read API — aggregate definition, policy, and data plane.
 */
import type { ApiResolvedPolicy } from "./policies"
import { apiClient, type ApiResult } from "./client"
import type { ApiAnnotationMode, ApiTask, ApiTaskStatus, ApiTaskType } from "./tasks"

export type ApiDataPointer = {
  id: string
  task_id: string
  location_ref: string
  access_policy_ref: string | null
  source_version_ref: string | null
  created_at: string
}

export type ApiTaskItemSetup = {
  id: string
  task_id: string
  data_pointer_id: string
  external_item_ref: string
  location_ref: string | null
  status: string
  payload_preview: Record<string, unknown>
  created_at: string
  updated_at: string | null
}

export type SetupRegistrationStatus = "pending_registration" | "partial" | "registered"

export type ApiTaskSetupDataPlane = {
  pointer_count: number
  item_count: number
  registration_status: SetupRegistrationStatus
  storage_provider_hint: string | null
}

export type ApiTaskSetup = {
  task: ApiTask
  resolved_policy: ApiResolvedPolicy
  data_pointers: ApiDataPointer[]
  task_items: ApiTaskItemSetup[]
  data_plane: ApiTaskSetupDataPlane
}

export function getTaskSetup(taskId: string): Promise<ApiResult<ApiTaskSetup>> {
  return apiClient.get<ApiTaskSetup>(`/tasks/${taskId}/setup`)
}

export function listTaskDataPointers(taskId: string): Promise<ApiResult<ApiDataPointer[]>> {
  return apiClient.get<ApiDataPointer[]>(`/tasks/${taskId}/data-pointers`)
}

export type { ApiAnnotationMode, ApiTaskStatus, ApiTaskType }

File: hej/apps/hej-api/app/schemas/task_setup.py
"""Read models for the task Setup surface."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.policies import ResolvedPolicyRead
from app.schemas.tasks import DataPointerRead, TaskItemRead, TaskRead

SetupRegistrationStatus = Literal["pending_registration", "partial", "registered"]


class TaskSetupDataPlaneRead(BaseModel):
    pointer_count: int
    item_count: int
    registration_status: SetupRegistrationStatus
    storage_provider_hint: str | None = None


class TaskSetupRead(BaseModel):
    task: TaskRead
    resolved_policy: ResolvedPolicyRead
    data_pointers: list[DataPointerRead] = Field(default_factory=list)
    task_items: list[TaskItemRead] = Field(default_factory=list)
    data_plane: TaskSetupDataPlaneRead
